[JIT] Python integration for custom operators

goldsborough opened the pull request:
Adds the Python path to custom operators, including dynamically loading operations into Python.

@zdevito

zdevito commented on torch/csrc/jit/init.cpp:
```diff
@@ -219,6 +223,24 @@ void initJITBindings(PyObject *module) {
+          "Found ", operations.size(), " overloads for operator ",
+          qualified_name, "! Overloads are not supported from Python.");
+      auto& op = operations[0];
+      return py::cpp_function([op](py::args args, py::kwargs kwargs){
```
nit: it is bit hard to see here that what we are caputuring is a `shared_ptr<Operator>`, might be clearer with
```
std::shared_ptr<Operator> op = operations[0];
```

zdevito commented on torch/csrc/jit/ivalue.h:
```diff
@@ -129,10 +135,12 @@ struct IValue {
   // While some of these accessors could be generated through templates,
   // we prefer to write them manually for clarity
 
+  Tag getTag() const noexcept { return tag; }
```
Don't expose the IValue tag. 
Also can you drop this comment in here:
```
// Note: IValue tags are intentionally private. In the future we may encode this value different (e.g. using NaN boxing), and this would make it more costly to determine the tag for all types vs just determining if something is a particular type. Instead we want clients to use the `isX` methods when possible. If for perf. reasons you really, absolutely, must have a jump table, then we can revisit this.

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -22,52 +26,143 @@ inline Stack createStack(const py::tuple& tuple, at::ArrayRef<Value*> inputs, si
+            [&key](const Argument& argument) { return argument.name == key; }),
+        "Unknown keyword argument '", key, "'");
+  }
+  // If there are unconsumed kwargs but none of them were uknown, the first
```
'uknown'

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -22,52 +26,143 @@ inline Stack createStack(const py::tuple& tuple, at::ArrayRef<Value*> inputs, si
+      args.size(), " arguments");
+
+  Stack stack;
+  stack.reserve(args.size() + kwargs.size());
```
`schema.arguments.size()` is more accurate for reserve since there may be default arguments.

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -22,52 +26,143 @@ inline Stack createStack(const py::tuple& tuple, at::ArrayRef<Value*> inputs, si
+  size_t consumed_kwargs = 0;
+  for (size_t i = args.size(); i < schema.arguments.size(); ++i) {
+    const auto& arg = schema.arguments[i];
+    if (kwargs.contains(arg.name.c_str())) {
```
Is there a way to avoid the double-hashing here?

goldsborough replied:
Looked, but can't see any convenient way in the pybind API ...

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -22,52 +26,143 @@ inline Stack createStack(const py::tuple& tuple, at::ArrayRef<Value*> inputs, si
+    const FunctionSchema& schema,
+    py::args args,
+    py::kwargs kwargs) {
+  AT_CHECK(
```
For these checks, it would be good to print the schema for context.

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -22,52 +26,143 @@ inline Stack createStack(const py::tuple& tuple, at::ArrayRef<Value*> inputs, si
+}
+
+// TODO: Remove once we clean up the GraphExecutor usage.
+inline Stack createStack(const py::tuple& tuple, at::ArrayRef<Value*> inputs, size_t reserve_extra_space = 0) {
```
Can we replace all easy-to-replace versions of createStack in this PR as well? For instance, Method has a `schema()` function so that should be easy to replace. Similar for wrapStack.

zdevito commented on torch/csrc/jit/python_custom_operator.cpp:
```diff
@@ -0,0 +1,48 @@
+    // If there is more than one return value, pop them into a py::tuple.
+    py::tuple return_values(stack.size());
+    for (size_t ret = 1; ret <= return_values.size(); ++ret) {
+      return_values[return_values.size() - ret] = toPyObject(pop(stack));
```
Argument and return values on the stack just appear in standard order:

```
c, d = foo(a,b)
before call: a b <- top of stack
after call: c, d <- top of stack
```
So this should just be `0..stack.size()`

zdevito reviewed (approved):
Look good. Minor fixes, and we should have tests, but otherwise ready to go.

goldsborough committed:
Python integration for custom ops

Robust error handling for invoke

Better error messages, reversed push order, dont expose IValue::Tag

Consolidate createStack/wrapStack methods

Wrote tests for dynamic binding and argument checking

goldsborough committed:
Undo invokeScriptMethodFromPython update

goldsborough committed:
Fix invokeScriptMethodFromPython

goldsborough committed:
Fix noexcept and add TORCH_API to getAllOperatorsFor
